File: acados_dc_motor/rosenbrock_m2bis.m
%%
clear; clc; close all;
%%

solver_dir = fullfile(pwd, 'c_generated_code');
% json_file  = fullfile(solver_dir, 'acados_ocp_generic_nlp.json');
json_file  = fullfile(pwd, 'acados_ocp_nlp.json');

solver_creation_opts = struct( ...
    'json_file', 'acados_ocp_nlp.json', ...
    'build', true, ...              % không build lại
    'generate', false, ...           % không render template lại
    'compile_mex_wrapper', false, ...% không compile mex lại
    'compile_interface', false, ...
    'output_dir', solver_dir);

% tạo solver từ json + code đã build sẵn
ocp_solver = AcadosOcpSolver([], solver_creation_opts);

%% solve the NLP
% initial guess
init_x = [2.5; 3.0];
ocp_solver.set('init_x', init_x);

% set the parameters
p_value = [1];
ocp_solver.set('p', p_value);

% solve and time
tic
ocp_solver.solve();
time_external = toc;
% internal timing
total_time = ocp_solver.get('time_tot');

% check status
status = ocp_solver.get('status');
if status ~= 0
    warning(['solver failed with status ',num2str(status)]);
end

% display results
x_opt = ocp_solver.get('x', 0);
disp('Optimal solution:')  % should be [1;1] for p = [1;1]
disp(x_opt)
disp(['Total time (internal): ', num2str(1e3*total_time), ' ms'])
disp(['Total time (external): ', num2str(1e3*time_external), ' ms'])

%
solving_time = [];
solving_internal_time = [];
xy_sol = [];
timer = tic;
for r_value=linspace(1,3,25)
    % set the parameters
    p_value = r_value;
    ocp_solver.set('p', p_value);
    timer_sol = tic;
    ocp_solver.solve();
    sol_t = toc(timer_sol);
    solving_time = [solving_time, sol_t*1e3];
    % internal timing
    total_internal_time = ocp_solver.get('time_tot');
    solving_internal_time = [solving_internal_time, total_internal_time*1e3]
    xy_sol = [xy_sol, ocp_solver.get('x', 0);]

end
total_time = toc(timer);  

File: acados_dc_motor/rosenbrock_m2bis.py
from acados_template import AcadosOcpSolver
import numpy as np
import casadi as ca

import time
from matplotlib import pyplot as plt

json_file = "acados_ocp_generic_nlp.json"
ocp_solver = AcadosOcpSolver(None, build=False, generate=False, json_file=json_file)

# warmup solver
ocp_solver.set(0, "x", np.array([0.0, 0.0]))
for k in range(5):
    r_example = 1.0 + 0.1 * k
    ocp_solver.set(0, "p", np.array([r_example]))
    status = ocp_solver.solve()

    if status != 0:
        print(f"warmup {k}, solver failed with returned status {status}")

r_values = np.linspace(0.0, 3.0, 101)

solving_time_ext = []
solving_time_int = []

xy_sol = []
cost_sol = []

t_init = time.time()
for r_value in r_values:
    ocp_solver.set(0, "p", np.array([r_value]))

    # solve
    start_time = time.time()
    status = ocp_solver.solve()
    solving_time = time.time() - start_time
    solving_time_ext.append(solving_time)

    x0 = ocp_solver.get(0, "x")
    xy_sol.append(x0)
    f_val = ocp_solver.get_cost()
    cost_sol.append(f_val)

    # print(f"r = {r_value:.2f}, x = {x0}, f = {f_val:.4f}, solve time = {solving_time*1e3:.2f} ms, status = {status}")

    t_int = ocp_solver.get_stats("time_tot")
    solving_time_int.append(t_int)

# total time
total_time = time.time() - t_init
print(f"Total time for {len(r_values)} solves: {total_time:.2f} s")

xy_sol = np.array(xy_sol)
cost_sol = np.array(cost_sol)
solving_time_ext = np.array(solving_time_ext)
solving_time_int = np.array(solving_time_int)

# plot results
plt.figure()
plt.subplot(3,1,1)
plt.plot(r_values, xy_sol[:,0], label='x1')
plt.plot(r_values, xy_sol[:,1], label='x2')
plt.xlabel('r')
plt.ylabel('Optimal x,y')

plt.subplot(3,1,2)
plt.plot(r_values, cost_sol)
plt.xlabel('r')
plt.ylabel('Optimal cost')

plt.subplot(3,1,3)
plt.plot(r_values, solving_time_ext*1e3, label='External time')
plt.plot(r_values, solving_time_int*1e3, label='Internal time')
plt.xlabel('r')
plt.ylabel('Solving time [ms]')
plt.legend()
plt.tight_layout()
plt.show()


